plugins/UserCountry/Commands/AttributeHistoricalDataWithLocations.php

```diff
@@ -0,0 +1,291 @@
+
+        $output->writeln(
+            sprintf('Re-attribution for date range: %s to %s. %d visits to process with provider "%s".',
+                $from, $to, $this->amountOfVisits, $this->visitorGeolocator->getProvider()->getId())
+        );
+
+        $this->timer = new Timer();
+
+        $this->processSpecifiedLogsInChunks($output, $from, $to, $input->getOption(self::SEGMENT_LIMIT_OPTION));
+
+        $output->writeln("Completed. <comment>" . $this->timer->__toString() . "</comment>");
+
+        return 0;
+    }
+
+    protected function processSpecifiedLogsInChunks(OutputInterface $output, $from, $to, $segmentLimit)
+    {
+        $visitFieldsToSelect = array_merge(array('idvisit', 'location_ip'), array_keys(self::$logVisitFieldsToUpdate));
+
+        $lastId = 0;
+        do {
+            $logs = $this->dao->getVisitsWithDatesLimit($from, $to, $visitFieldsToSelect, $lastId, $segmentLimit);
+            if (!empty($logs)) {
+                $lastId = $logs[count($logs) - 1]['idvisit'];
+
+                $this->reattributeVisitLogs($output, $logs);
+            }
+        } while (count($logs) == $segmentLimit);
+    }
+
+    protected function reattributeVisitLogs(OutputInterface $output, $logRows)
+    {
+        foreach ($logRows as $row) {
+            if (!$this->isRowComplete($output, $row)) {
+                continue;
+            }
+
+            $location = $this->getVisitLocation($row);
+            $valuesToUpdate = $this->getValuesToUpdate($row, $location);
+
+            $this->onVisitProcessed($output);
+
+            $idVisit = $row['idvisit'];
+
+            if (empty($valuesToUpdate)) {
+                $this->writeIfVerbose($output, 'Nothing to update for idvisit = ' . $idVisit . '. Existing location info is same as geolocated.');
+            } else {
+                $this->writeIfVerbose($output, 'Updating visit with idvisit = ' . $idVisit . '.');
+
+                $this->dao->updateVisits($valuesToUpdate, $idVisit);
+                $this->dao->updateConversions($valuesToUpdate, $idVisit);
+            }
+        }
+    }
+
+    /**
+     * @param InputInterface $input
+     * @return int
+     */
+    protected function getPercentStep(InputInterface $input)
+    {
+        $percentStep = $input->getArgument(self::PERCENT_STEP_ARGUMENT);
+
+        if (!is_numeric($percentStep)) {
+            return self::PERCENT_STEP_ARGUMENT_DEFAULT;
+        }
+
+        // Percent step should be between maximum percent value and minimum percent value (1-100)
+        if ($percentStep > 99 || $percentStep < 1) {
+            return 100;
+        }
+
+        return $percentStep;
+    }
+
+    /**
+     * Returns location log values that are different than the values currently in a log row.
+     *
+     * @param array $row The visit row.
+     * @param array $location The location information.
+     * @return array The location properties to update.
+     */
+    protected function getValuesToUpdate(array $row, $location)
```

Comment:
> I don't think it would suggest that all callers should use this method.

It would make the strategy used canon, so-to-speak, which I think is a bad idea. Deviations would create code redundancy (ie, people reimplementing the use case in another way), which would lead to future issues w/ code reviews.

> It's basically a Controller and shouldn't have any logic etc. It should only get some input params, trigger some action/business logic and care about output).

I understand what you're getting at, that the Command classes are really just gateways to application logic, rather than the containers of the logic themselves. However, I disagree that they are the same as Controllers.

Controllers provide access to short to medium timed functionality where the output is returned all at once, whereas commands provide access to short to very long running functionality where output/feedback is provided to the user in real-time, some of which can be 'verbose' which doesn't necessarily mean debug output. This difference in requirements makes the comparison invalid; commands are not just gateways to application logic, they are containers of custom logic that glues different pieces of application logic together.

I think the issue w/ command testability isn't really the fault of putting logic in command classes, but the inferior design of Symfony's Command base. Using the class requires ONLY being able to use it within Symfony's console command framework. You can't make `execute(...)` public. You can't define public methods meant for re-use because it would look strange. Ideally, the command logic & the command metadata (ie, the argument/option metadata, the logic reading from $input, writing to $output) should be separated automatically by the command line framework. Perhaps this can be added to core\Console.php...

> If an iterator is the best solution to solve it, use it ;)

The iterator would be added to core (since it would traverse the log_visit table or since it might be implemented as an adapter), and thus might cause political problems. Maybe better to do in a followup issue.